### PR TITLE
Index selection

### LINKED ISSUE #1959: add index selection keyword
### Expected behaviour
The [selection language](https://www.mdanalysis.org/docs/documentation_pages/selections.html) should allow me to use the output from `AtomGroup.indices` for selections, which are the *0-based indices*.

### Actual behaviour
Only [`bynum`](https://www.mdanalysis.org/docs/documentation_pages/selections.html#index) is implemented as a selection keyword for indices. It uses *1-based* indices. 

### Code to reproduce the behaviour
See https://github.com/MDAnalysis/mdanalysis/issues/1958#issuecomment-400804285.

### Proposed solution
Implement a new selection keyword `index` that does the same as `bynum` but for 0-based indices.

### Currently version of MDAnalysis:
(run `python -c "import MDAnalysis as mda; print(mda.__version__)"`)
develop

### PR DESCRIPTION
**Pull request by IAlibay**
Fixes #1959 

Changes made in this Pull Request:
 - Adds new selection keyword `index`. This has the exact same behaviour as `bynum` but is 0-based rather than 1-based.

Notes:
 - I might have gone a little bit overboard in terms of duplicating the `bynum` tests for `index` in `test_atomselections`. If it's too much I can revert back to only including `test_index`.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: testsuite/MDAnalysisTests/core/test_atomselections.py [hunk 1]
```diff
@@ -392,25 +420,32 @@ def test_resid_single(self, universe):
...
+            sel_string = 'cylayer 10 20 20 -20 ' + atom
+            sel = atgp.select_atoms(sel_string)
+            assert_equal(len(sel), 1155)
 
     def test_cyzone(self, universe):
         """Cylinder zone selections with tricilinic periodicity (Issue 274)"""
         atgp = universe.select_atoms('name OW')
-        sel = atgp.select_atoms('cyzone 20 20 -20 bynum 3554')
-        assert_equal(len(sel), 1556)
+        for atom in ['bynum 3554', 'index 3553']:
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/core/test_atomselections.py]: This sort of thing is best done with fixtures

File: testsuite/MDAnalysisTests/core/test_atomselections.py [hunk 2]
```diff
@@ -392,25 +420,32 @@ def test_resid_single(self, universe):
 
     def test_same_coordinate(self, universe):
         """Test the 'same ... as' construct (Issue 217)"""
-        sel = universe.select_atoms("same x as bynum 1 or bynum 10")
-        assert_equal(len(sel), 12,
-                     "Found a wrong number of atoms with same x as ids 1 or 10")
-        target_ids = np.array([0, 8, 9, 224, 643, 3515,
-                               11210, 14121, 18430, 25418, 35811, 43618])
-        assert_equal(sel.indices, target_ids,
-                     "Found wrong atoms with same x as ids 1 or 10")
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/core/test_atomselections.py]: Does the implicit OR work here? Ie “index 0 9”

Comment (IAlibay) [reply testsuite/MDAnalysisTests/core/test_atomselections.py]: It does, I've added those cases to the list of strings being tested, and also added index to `test_range_selections` in `TestImplicitOr`.

**Comment by IAlibay** [comment]
Thanks for the comments @richardjgowers. Unfortunately I am not very versed in pytest fixtures, I've changed the `for` constructs for `pytest.mark.parametrize`. If you had something else in mind, I will gladly change this though.

**Comment by richardjgowers** [comment]
@IAlibay yeah looks good.  The annoying thing with tests is that they will fail on the first thing wrong, so if you've got lots of things in one test method, you'll only know the first failure.  If you instead parametrize you get more detail about how/why it's failing which helps with debugging.

**Comment by orbeckst** [comment]
Thank you both!

**Comment by orbeckst** [comment]
Thank you both!

**Pull request merged**